--- /home/antonio/projects/bitbucket/paragon-ntfs-driver-porting-development/assets/tmp/paragon/configure	2017-09-08 18:06:41.000000000 +0200
+++ /home/antonio/projects/bitbucket/paragon-ntfs-driver-porting-development/configure	2018-12-28 15:30:56.916867390 +0100
@@ -4659,6 +4659,112 @@
 fi
 
 #Check members
+
+# <Antonio-Petricca@2018>
+
+########
+# 4.15 #
+########
+
+ac_fn_c_check_decl "$LINENO" "global_page_state" "ac_cv_have_decl_global_page_state" "#include <linux/swap.h>
+"
+if test "x$ac_cv_have_decl_global_page_state" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_DECL_GLOBAL_PAGE_STATE $ac_have_decl
+_ACEOF
+
+ac_fn_c_check_decl "$LINENO" "setup_timer" "ac_cv_have_decl_setup_timer" "#include <linux/timer.h>
+"
+if test "x$ac_cv_have_decl_setup_timer" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_DECL_SETUP_TIMER $ac_have_decl
+_ACEOF
+
+ac_fn_c_check_member "$LINENO" "struct bio" "bi_bdev" "ac_cv_member_struct_bio_bi_bdev" "#include <linux/blk_types.h>
+"
+if test "x$ac_cv_member_struct_bio_bi_bdev" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_STRUCT_BIO_BI_BDEV $ac_have_decl
+_ACEOF
+
+ac_fn_c_check_member "$LINENO" "struct bio" "bi_error" "ac_cv_member_struct_bio_bi_error" "#include <linux/blk_types.h>
+"
+if test "x$ac_cv_member_struct_bio_bi_error" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_STRUCT_BIO_BI_ERROR $ac_have_decl
+_ACEOF
+
+########
+# 4.16 #
+########
+
+ac_fn_c_check_decl "$LINENO" "init_buffer" "ac_cv_have_decl_init_buffer" "#include <linux/buffer_head.h>
+"
+
+if test "x$ac_cv_have_decl_init_buffer" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_DECL_INIT_BUFFER $ac_have_decl
+_ACEOF
+
+########
+# 4.17 #
+########
+
+ac_fn_c_check_member "$LINENO" "struct address_space" "tree_lock" "ac_cv_member_struct_address_space_tree_lock" "#include <linux/fs.h>
+"
+if test "x$ac_cv_member_struct_address_space_tree_lock" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_STRUCT_ADDRESS_SPACE_TREE_LOCK $ac_have_decl
+_ACEOF
+
+########
+# 4.20 #
+########
+
+ac_fn_c_check_decl "$LINENO" "time_to_tm" "ac_cv_have_decl_time_to_tm" "#include <linux/time32.h>
+"
+if test "x$ac_cv_have_decl_time_to_tm" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_DECL_TIME_TO_TM $ac_have_decl
+_ACEOF
+
+
+
+#######
+# 5.0 #
+#######
+
+ac_fn_c_check_decl "$LINENO" "current_kernel_time64" "ac_cv_have_decl_current_kernel_time64" "#include <linux/ktime.h>
+"
+
+if test "x$ac_cv_have_decl_current_kernel_time64" = xyes; then :
+  ac_have_decl=1
+else
+  ac_have_decl=0
+fi
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_DECL_CURRENT_KERNEL_TIME64 $ac_have_decl
+_ACEOF
+
+ac_have_decl=0
+
+for ac_header in uapi/linux/mount.h
+do :
+  ac_fn_c_check_header_compile "$LINENO" "uapi/linux/mount.h" "ac_cv_header_uapi_linux_mount_h" ""
+if test "x$ac_cv_header_uapi_linux_mount_h" = xyes; then :
+  ac_have_decl=1
+fi
+
+done
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_UAPI_LINUX_MOUNT_H $ac_have_decl
+_ACEOF
+
+# </Antonio-Petricca@2019>
+
 ac_fn_c_check_member "$LINENO" "struct bio" "bi_iter" "ac_cv_member_struct_bio_bi_iter" "#include <linux/blk_types.h>
 "
 if test "x$ac_cv_member_struct_bio_bi_iter" = xyes; then :
