--- /home/antonio/projects/bitbucket/paragon-ntfs-driver-porting-development/assets/tmp/paragon/ifslinux/ufsdvfs.c	2017-09-08 18:06:41.000000000 +0200
+++ /home/antonio/projects/bitbucket/paragon-ntfs-driver-porting-development/ifslinux/ufsdvfs.c	2019-01-09 06:44:03.368958631 +0100
@@ -707,10 +707,21 @@
 //
 // Converts posix time into timestamp
 ///////////////////////////////////////////////////////////
+
 static inline void
 posix2kernel(
     IN  UINT64  tm,
+// <Antonio-Petricca@2018>
+
+  // OUT struct timespec *ts
+
+#if IS_TIMESPEC_64_BIT
     OUT struct timespec *ts
+#else
+    OUT struct timespec64 *ts
+#endif
+
+// </Antonio-Petricca@2018>
     )
 {
   union utimespec ut;
@@ -719,7 +730,6 @@
   ts->tv_nsec = ut.tv_nsec;
 }
 
-
 ///////////////////////////////////////////////////////////
 // kernel2posix
 //
@@ -727,7 +737,17 @@
 ///////////////////////////////////////////////////////////
 static inline UINT64
 kernel2posix(
-    IN const struct timespec *ts
+// <Antonio-Petricca@2018>
+
+  // IN struct timespec *ts
+
+#if IS_TIMESPEC_64_BIT
+    IN struct timespec *ts
+#else
+    IN struct timespec64 *ts
+#endif
+
+// </Antonio-Petricca@2018>
     )
 {
   union utimespec ut;
@@ -746,7 +766,29 @@
 UINT64 UFSDAPI_CALL
 current_time_posix( void )
 {
-  struct timespec ts = current_kernel_time();
+// <Antonio-Petricca@2018>
+
+  //struct timespec ts = current_kernel_time();
+
+#if IS_TIMESPEC_64_BIT
+    struct timespec ts = current_kernel_time();
+#else
+    // <Antonio-Petricca@2019>
+
+    //struct timespec64 ts = current_kernel_time64();
+
+    #if HAVE_DECL_CURRENT_KERNEL_TIME64
+      struct timespec64 ts = current_kernel_time64();
+    #else
+      struct timespec64 ts;
+      ktime_get_coarse_real_ts64(&ts);
+    #endif
+
+    // </Antonio-Petricca@2019>
+#endif
+
+// </Antonio-Petricca@2018>
+
   return kernel2posix( &ts );
 }
 #else
@@ -766,7 +808,17 @@
 ///////////////////////////////////////////////////////////
 static inline UINT64
 kernel2nt(
-    IN const struct timespec *ts
+// <Antonio-Petricca@2018>
+
+  // IN struct timespec *ts
+
+#if IS_TIMESPEC_64_BIT
+    IN struct timespec *ts
+#else
+    IN struct timespec64 *ts
+#endif
+
+// </Antonio-Petricca@2018>
     )
 {
   // 10^7 units of 100 nanoseconds in one second
@@ -782,7 +834,17 @@
 static inline void
 nt2kernel(
     IN  const UINT64    tm,
+// <Antonio-Petricca@2018>
+
+  // OUT struct timespec *ts
+
+#if IS_TIMESPEC_64_BIT
     OUT struct timespec *ts
+#else
+    OUT struct timespec64 *ts
+#endif
+
+// </Antonio-Petricca@2018>
     )
 {
   UINT64 t    = tm - _100ns2seconds*SecondsToStartOf1970;
@@ -800,7 +862,29 @@
 UINT64 UFSDAPI_CALL
 current_time_nt( void )
 {
-  struct timespec ts = current_kernel_time();
+// <Antonio-Petricca@2018>
+
+  //struct timespec ts = current_kernel_time();
+
+#if IS_TIMESPEC_64_BIT
+    struct timespec ts = current_kernel_time();
+#else
+    // <Antonio-Petricca@2019>
+
+    //struct timespec64 ts = current_kernel_time64();
+
+    #if HAVE_DECL_CURRENT_KERNEL_TIME64
+      struct timespec64 ts = current_kernel_time64();
+    #else
+      struct timespec64 ts;
+      ktime_get_coarse_real_ts64(&ts);
+    #endif
+
+    // </Antonio-Petricca@2019>
+#endif
+
+// </Antonio-Petricca@2018>
+
   return kernel2nt( &ts );
 }
 
@@ -815,12 +899,45 @@
 //
 // Returns current time (to store in inode)
 ///////////////////////////////////////////////////////////
-static inline struct timespec
+// <Antonio-Petricca@2018>
+
+  //static inline struct timespec
+
+#if IS_TIMESPEC_64_BIT
+    static inline struct timespec
+#else
+    static inline struct timespec64
+#endif
+
+// </Antonio-Petricca@2018>
+
 ufsd_inode_current_time(
     IN usuper *sbi
     )
 {
-  struct timespec ts = current_kernel_time();
+// <Antonio-Petricca@2018>
+
+  //struct timespec ts = current_kernel_time();
+
+#if IS_TIMESPEC_64_BIT
+    struct timespec ts = current_kernel_time();
+#else
+    // <Antonio-Petricca@2019>
+
+    //struct timespec64 ts = current_kernel_time64();
+
+    #if HAVE_DECL_CURRENT_KERNEL_TIME64
+      struct timespec64 ts = current_kernel_time64();
+    #else
+      struct timespec64 ts;
+      ktime_get_coarse_real_ts64(&ts);
+    #endif
+
+    // </Antonio-Petricca@2019>
+#endif
+
+// </Antonio-Petricca@2018>
+
   if ( is_hfs( &sbi->options ) )
     ts.tv_nsec = 0;
   else if ( is_fat( &sbi->options ) ) {
@@ -845,14 +962,39 @@
 // Truncate time to a granularity
 // Returns 1 if changed
 ///////////////////////////////////////////////////////////
+
 static inline int
 ufsd_time_trunc(
     IN usuper *sbi,
+// <Antonio-Petricca@2018>
+
+/*
     IN const struct timespec *ts,
     IN OUT struct timespec *td
     )
 {
   struct timespec t;
+*/
+
+#if IS_TIMESPEC_64_BIT
+
+    IN const struct timespec *ts,
+    IN OUT struct timespec *td
+    )
+{
+  struct timespec t;
+
+#else
+
+    IN const struct timespec64 *ts,
+    IN OUT struct timespec64 *td
+    )
+{
+  struct timespec64 t;
+
+#endif
+
+// </Antonio-Petricca@2018>
 
   t.tv_sec  = ts->tv_sec;
 
@@ -890,7 +1032,17 @@
 static inline UINT64
 kernel2ufsd(
     IN const usuper *sbi,
-    IN const struct timespec *ts
+// <Antonio-Petricca@2018>
+
+  // IN struct timespec *ts
+
+#if IS_TIMESPEC_64_BIT
+    IN struct timespec *ts
+#else
+    IN struct timespec64 *ts
+#endif
+
+// </Antonio-Petricca@2018>
     )
 {
   return is_posixtime( &sbi->options )? kernel2posix( ts ) : kernel2nt( ts );
@@ -1233,7 +1385,9 @@
     DebugTrace( 0, UFSD_LEVEL_ERROR, ("**** seq=%x: size 0x%x  asize 0x%x", block->seq, block->size, block->asize ));
     DebugTrace( 0, UFSD_LEVEL_ERROR, ("**** HeapFree(%p) %s barrier failed at 0x%zx", p, hint, PtrOffset( block, tst ) + o ));
     ufsd_turn_on_trace_level();
+
     ufsdapi_dump_memory( block, 512 );
+
     ufsd_revert_trace_level();
     BUG_ON(1);
   }
@@ -1704,7 +1858,26 @@
   }
 
   bh->b_state = 0;
+
+// <Antonio-Petricca@2018>
+
+  // init_buffer( bh, end_buffer_read_sync, NULL );
+
+#if HAVE_DECL_INIT_BUFFER 
+
   init_buffer( bh, end_buffer_read_sync, NULL );
+
+#else
+
+  // https://elixir.bootlin.com/linux/v4.13.12/source/fs/buffer.c
+
+  bh->b_end_io = end_buffer_read_sync;
+	bh->b_private = NULL;
+
+#endif
+
+ // </Antonio-Petricca@2018>
+
   atomic_set( &bh->b_count, 2 );
   set_bh_page( bh, page, bytes2skip );
   bh->b_size    = 512;
@@ -1776,8 +1949,21 @@
   unsigned long nr_pages, nr_anon, nr_free, max_ra;
   struct blk_plug plug;
 
-  nr_anon = global_page_state( NR_ACTIVE_ANON );
-  nr_free = global_page_state( NR_FREE_PAGES );
+  // <Antonio-Petricca@2018>
+
+  //nr_anon = global_page_state( NR_ACTIVE_ANON );
+  //nr_free = global_page_state( NR_FREE_PAGES );
+
+  #if !HAVE_DECL_GLOBAL_PAGE_STATE
+    nr_anon = global_zone_page_state( NR_ACTIVE_ANON );
+    nr_free = global_zone_page_state( NR_FREE_PAGES );
+  #else
+    nr_anon = global_page_state( NR_ACTIVE_ANON );
+    nr_free = global_page_state( NR_FREE_PAGES );
+  #endif
+
+  // </Antonio-Petricca@2018>
+
   max_ra  = ( nr_anon + nr_free ) >> 1;
 
   if ( 0 != sbi->options.raKb ) {
@@ -1809,10 +1995,28 @@
   for ( nr_pages = 0; start < end; start++ ) {
     struct page *page;
 
+// <Antonio-Petricca@2018>
+
+    /*spin_lock_irq( &mapping->tree_lock );
+    page = radix_tree_lookup( &mapping->page_tree, start );
+    spin_unlock_irq( &mapping->tree_lock );*/
+
+#if HAVE_STRUCT_ADDRESS_SPACE_TREE_LOCK
+
     spin_lock_irq( &mapping->tree_lock );
     page = radix_tree_lookup( &mapping->page_tree, start );
     spin_unlock_irq( &mapping->tree_lock );
 
+#else
+
+    spin_lock_irq( &mapping->i_pages.xa_lock );
+    page = radix_tree_lookup( &mapping->i_pages, start );
+    spin_unlock_irq( &mapping->i_pages.xa_lock );
+
+#endif
+
+// </Antonio-Petricca@2018>
+
 #if is_decl( RADIX_TREE_EXCEPTIONAL_ENTRY )
     if ( NULL != page && !radix_tree_exceptional_entry( page ) )
       continue;
@@ -1885,7 +2089,18 @@
   sector_t  devblock        = offset >> sb->s_blocksize_bits;
   unsigned long nBlocks     = bytes >> sb->s_blocksize_bits;
   unsigned long cnt         = 0;
-  unsigned long limit       = global_page_state( NR_FREE_PAGES ) << (PAGE_SHIFT - sb->s_blocksize_bits);
+
+  // <Antonio-Petricca@2018>
+
+  //unsigned long limit = global_page_state( NR_FREE_PAGES ) << (PAGE_SHIFT - sb->s_blocksize_bits);
+
+  #if !HAVE_DECL_GLOBAL_PAGE_STATE
+    unsigned long limit = global_zone_page_state( NR_FREE_PAGES ) << (PAGE_SHIFT - sb->s_blocksize_bits);
+  #else
+    unsigned long limit = global_page_state( NR_FREE_PAGES ) << (PAGE_SHIFT - sb->s_blocksize_bits);
+  #endif
+
+  // </Antonio-Petricca@2018>
 
   if ( limit >= 0x2000 )
     limit -= 0x1000;
@@ -5493,8 +5708,24 @@
 {
   if ( NULL == acl )
     return;
+
+// <Antonio-Petricca@2018>
+
+  /*
+    if ( atomic_dec_and_test( &acl->a_refcount ) )
+      kfree( acl );
+  */
+
+#if IS_STRUCT_POSIXACL_AREFCOUNT_OF_TYPE_ATOMICT
   if ( atomic_dec_and_test( &acl->a_refcount ) )
     kfree( acl );
+#else
+  if ( atomic_dec_and_test( &acl->a_refcount.refs ) )
+    kfree( acl );
+#endif
+
+// </Antonio-Petricca@2018>
+
 }
 
 
@@ -7045,7 +7276,17 @@
 {
   ssize_t ret;
   loff_t  end, i_size;
-  struct timespec now;
+// <Antonio-Petricca@2018>
+
+  // struct timespec now;
+
+#if IS_TIMESPEC_64_BIT
+    struct timespec now;
+#else
+    struct timespec64 now;
+#endif
+
+// </Antonio-Petricca@2018>
   int dirty                     = 0;
   struct file  *file            = iocb->ki_filp;
   struct address_space *mapping = file->f_mapping;
@@ -8612,8 +8853,20 @@
 #ifdef BIO_UPTODATE
   err = !test_bit( BIO_UPTODATE, &bio->bi_flags );
 #else
+  // <Antonio-Petricca@2018>
+
   // 4.3+
-  int err = bio->bi_error;
+  //int err = bio->bi_error;
+
+  #if !HAVE_STRUCT_BIO_BI_ERROR
+    int err = (BLK_STS_OK != bio->bi_status);
+
+  #else
+    // 4.3+
+    int err = bio->bi_error;
+  #endif
+
+  // </Antonio-Petricca@2018>
 #endif
 
 //  printk( "end_io_read at %llx sz=%x, cnt=%x\n", (UINT64)BIO_BISECTOR( bio ) << 9, BIO_BISIZE( bio ), (unsigned)bio->bi_vcnt );
@@ -8663,8 +8916,20 @@
 #ifdef BIO_UPTODATE
   err = !test_bit( BIO_UPTODATE, &bio->bi_flags );
 #else
+  // <Antonio-Petricca@2018>
+
   // 4.3+
-  int err = bio->bi_error;
+  //int err = bio->bi_error;
+
+  #if !HAVE_STRUCT_BIO_BI_ERROR
+    int err = (BLK_STS_OK != bio->bi_status);
+
+  #else
+    // 4.3+
+    int err = bio->bi_error;
+  #endif
+
+  // </Antonio-Petricca@2018>
 #endif
 
   if ( err ) {
@@ -8704,7 +8969,19 @@
     struct bio *bio = bio_alloc( GFP_NOFS|__GFP_HIGH, nr_vecs ); // GFP_NOIO
     if ( likely( NULL != bio ) ) {
       BIO_BISECTOR( bio ) = first_sector;
-      bio->bi_bdev    = bdev;
+
+      // <Antonio-Petricca@2018>
+
+      //bio->bi_bdev    = bdev;
+
+      #if !HAVE_STRUCT_BIO_BI_BDEV  
+        // https://lkml.org/lkml/2017/8/25/136
+        bio_set_dev(bio, bdev);
+      #else
+        bio->bi_bdev = bdev;
+      #endif
+
+      // </Antonio-Petricca@2018>
 
       DebugTrace( 0, UFSD_LEVEL_BIO, ("bio+: o=%" PSCT "x", first_sector << 9 ));
       return bio;
@@ -12366,18 +12643,29 @@
   mod_timer( &sbi->flush_timer, HZ + sbi->last_dirty + msecs_to_jiffies( sec * 1000 ) );
 }
 
-
 ///////////////////////////////////////////////////////////
 // flush_timer_fn
 //
 // Timer function
 ///////////////////////////////////////////////////////////
-static void
-flush_timer_fn(
-    IN unsigned long data
-    )
-{
-  usuper *sbi = (usuper*)data;
+
+// <Antonio-Petricca@2018>
+
+#if !HAVE_DECL_SETUP_TIMER
+
+  static void flush_timer_fn(struct timer_list *t)
+  {
+    usuper* sbi = from_timer(sbi, t, flush_timer);
+  
+#else
+
+  static void flush_timer_fn(IN unsigned long data)
+  {
+    usuper *sbi = (usuper*)data;
+
+#endif
+
+// </Antonio-Petricca@2018>
 
   if ( !sbi->bdirty ) {
     // Do not wake up flush thread
@@ -12423,7 +12711,18 @@
   //
   // Set up an interval timer which can be used to trigger a flush wakeup after the flush interval expires
   //
+
+// <Antonio-Petricca@2018>
+  
+  // setup_timer( &sbi->flush_timer, flush_timer_fn, (unsigned long)sbi );
+
+#if HAVE_DECL_SETUP_TIMER
   setup_timer( &sbi->flush_timer, flush_timer_fn, (unsigned long)sbi );
+#else
+  timer_setup( &sbi->flush_timer, flush_timer_fn, 0 );
+#endif
+
+// </Antonio-Petricca@2018>
 
   wake_up( &sbi->wait_done_flush );
 
@@ -14294,6 +14593,11 @@
       "Default options: " DEFAULT_MOUNT_OPTIONS "\n"
 #endif
       "" );
+
+// <Antonio-Petricca@2018>
+  printk( KERN_NOTICE QUOTED_UFSD_DEVICE": ported on kernels 4.13 to 5.0 by Antonio Petricca @ 2018-2019 (https://github.com/antonio-petricca/paragon-ufsd-ntfs-driver-porting).");
+// </Antonio-Petricca@2018>
+
   printk( KERN_NOTICE QUOTED_UFSD_DEVICE": PAGE_SIZE=%uK, THREAD_SIZE=%uk\n", (unsigned)(PAGE_SIZE>>10), (unsigned)(THREAD_SIZE>>10) );
 #endif
 
@@ -14431,7 +14735,13 @@
 //
 MODULE_DESCRIPTION("Paragon " QUOTED_UFSD_DEVICE " driver");
 MODULE_AUTHOR("Andrey Shedel & Alexander Mamaev");
-MODULE_LICENSE("Commercial product");
+
+// <Antonio-Petricca@2018>
+
+//MODULE_LICENSE("Commercial product");
+MODULE_LICENSE("GPL"); // Allow call of "vfs_write"
+
+// </Antonio-Petricca@2018>
 
 #ifdef UFSD_TRACE
 module_param_string(trace, ufsd_trace_level_, sizeof(ufsd_trace_level_), S_IRUGO);
